File: flyweight/PlayersFactory.java
package flyweight;

import java.util.HashMap;
import java.util.Map;

public class PlayersFactory {

    private static final Map<PlayerType, Player> inMemoryCachedPlayers = new HashMap<>();

    private PlayersFactory(){
    }

    public static Player getPlayer(PlayerType playerType) {
        if (inMemoryCachedPlayers.containsKey(playerType)) {
            return inMemoryCachedPlayers.get(playerType);
        }
        Player player = null;
        switch (playerType) {
            case MAIN_PLAYER -> player = new MainPlayer("Default Main Player");
            case WEAK_ENEMY -> player = new Enemy(1, 10);
            case STRONG_ENEMY -> player = new Enemy(5, 50);
        }
        if (player == null) {
            throw new UnsupportedOperationException("Unsupported player type");
        }
        inMemoryCachedPlayers.put(playerType, player);
        return player;
    }
}

File: C#/Flyweight/PlayersFactory.cs
using System.Diagnostics;

namespace Flyweight;

public sealed class PlayersFactory
{
    private static readonly Dictionary<PlayerType, IPlayer> _inMemoryCachedPlayers = [];

    public static IPlayer GetPlayer(PlayerType playerType)
    {
        if (_inMemoryCachedPlayers.TryGetValue(playerType, out IPlayer? value))
        {
            return value;
        }

        IPlayer player = playerType switch
        {
            PlayerType.MAIN_PLAYER => new MainPlayer("Default Main Player"),
            PlayerType.WEAK_ENEMY => new Enemy(1, 50),
            PlayerType.STRONG_ENEMY => new Enemy(5, 50),
            _ => throw new UnreachableException("Unsupported player type")
        };

        _inMemoryCachedPlayers.Add(playerType, player);
        return player;
    }
}
